reana_job_controller/job.py

```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of REANA.
+# Copyright (C) 2019 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Job interface for job management."""
+
+
+class Job():
+    """Job management interface."""
+
+    def __init__(self, docker_img, cmd, env_vars):
+        """Instanciates basic job.
+
+        :param docker_img: Docker image.
+        :type docker_img: str
+        :param cmd: Command to execute.
+        :type cmd: str
+        :param env_vars: Environment variables.
+        :type env_vars: dict
+        """
+        self.docker_img = docker_img
+        self.cmd = cmd
+        self.env_vars = env_vars
+
+    def submission_hook(fn):
+        """Add before and after submission hooks."""
+        def wrapper(inst, *args, **kwargs):
+            inst.before_submission()
+            result = fn(inst, *args, **kwargs)
+            inst.after_submission()
+            return result
+        return wrapper
+
+    def before_submission(self):
```

Comment:
`s/submission/execution/g` everywhere? Seems more natural to take the r-j-controller's internal point of view and talk about *executing* jobs that were "already submitted" to it by the r-w-engine. E.g. thinking of k8s or docker execution scenarios irst.  For HPC/HTC scenarios, the jobs will be sent (i.e. submitted) elsewhere indeed, so e.g. from SLURM's point of view there will be another submission, however from the r-j-c's internal point of view this is just a way of executing the jobs it was asked to run; the r-j-c will just select to proxy them to another backend.  (In other words, a "primary submission" is happening between r-w-e and r-j-c, and it is only a "secondary submission" that is happening between r-j-c and SLURM for some backends, so to speak.) Hence my preference for speaking about executing rather then submitting jobs inside r-j-c in general.